### ISSUE TITLE
Bulk insertion of nodes

### ISSUE THREAD

**Issue by cassidyclawson** [neo4jrb/neo4j#1079]
Hello!

Thanks everyone for the hardwork on this project. It's incredible.

Apologies if this has been covered. I've searched but couldn't find anything.

I need to MERGE up to approximately 1000 nodes and create 1000 relations in a single (preferably) synchronous web request. What is the most performant method of doing this?

I know I can use property mapping to pass in an array of properties for bulk creation with Cypher. The documentation doesn't mention this working with merge.

Can I do this with Neo4j.rb? If not, what is the best technique in the context of a Rails app / synchronous request. Thanks much!

**Comment by cheerfulstoic** [comment]
Thanks!

I would love to make batch queries work with Neo4j.  I've been slowly working on a new session API to make this possible.  In the meantime you can make your own HTTP calls to the Neo4j transaction API with a set of Cypher queries that you would like to execute.  On my phone so I can't lookup right now.

If you want to try the new session API which allows for batch Cypher queries it was released in 6.0.0 of neo4j-core but not really documented.  It's under the Neo4j::Core::CypherSession namespace if you want to take a peek. I can give some more info later when I'm at my computer.

Brian
;p

**Comment by cheerfulstoic** [comment]
Here's the Neo4j documentation on the REST JSON endpoints for transactions where you can send multiple Cypher queries at once:

http://neo4j.com/docs/stable/rest-api-transactional.html#rest-api-begin-and-commit-a-transaction-in-one-request

Here's a schema diagram of the new session API:

http://alpha264.imgur.com/all/

It uses adaptors for executing Cypher statements through HTTP and Embedded (Java/jRuby) mode.  Neo4j 3.0 will be coming out at some point with support for their new Bolt protocol (socket-based binary protocol) and when that comes we'll have an adaptor for that.

I also should note that there's a gem called neo4apis:

https://github.com/neo4jrb/neo4apis

It is basically a DSL that lets you set up a class which can take objects from one data source and import them into Neo4j.  It automatically batches the Cypher queries and executes them in batches for you.  I've made a few gems which use this.  You might check out `neo4apis-twitter` for an additional example of how it's used.

**Comment by cassidyclawson** [comment]
Hi Brian,

Thank you so much for this thoughtful response! I'll look through the docs and make a plan. 

I've toyed with activerecord-import (https://github.com/zdennis/activerecord-import) which has a pretty nice API for bulk record creation for AR. 

```
columns = [ :title, :author ]
values = [ ['Book1', 'FooManChu'], ['Book2', 'Bob Jones'] ]
Book.import columns, values
```

Something similar for Neo4j.rb would be epic. I would think my use case is exceedingly common because construction of a social graph will often entail an initial dump of contacts and relationships when a user joins the network. Look forward to seeing your progress!

**Comment by cheerfulstoic** [comment]
Yeah, `activerecord-import` is great.  I also hear that the `seqel` gem is good too, though I haven't gotten a chance to use it.

I should also point out this gem in case you have any SQL databases that you'd like to try copying to Neo4j (especially but not necessarily if you use ActiveRecord with them):

https://github.com/neo4jrb/neo4apis-activerecord

That gem has a couple of bugs and I've been considering re-writing it or taking another approach, but it's worth trying.

Regarding mass import of rows into Neo4j, I think that an `activerecord-import` alternative to Neo4j could be really cool.  Though I think there's been less of a movement toward it because of the `LOAD CSV` command.  Also there's the `neo4j-import` command, though that requires that you put your data into a specific CSV format and that you must be creating a fresh DB, but it's advantage is that you can get imports on the order of 1 million rows / second.

What might be nice, honestly, is some sort of tool that can automatically import your data via `LOAD CSV` or `neo4j-import`.  If you had your data in CSV or JSON, for example, you could define some mappings and it would take care of generating the Cypher `LOAD CSV` queries or the files for `neo4j-import`, depending on what you need.
